lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs

```diff
@@ -0,0 +1,184 @@
+using Org.Apache.REEF.Utilities.Diagnostics;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.IMRU.OnREEF.Driver
+{
+    /// <summary>
+    /// Class that handles failed evaluators and also provides Service 
+    /// and Context configuration
+    /// </summary>
+    /// <typeparam name="TMapInput"></typeparam>
+    /// <typeparam name="TMapOutput"></typeparam>
+    internal class ServiceAndContextConfigurationProvider<TMapInput,TMapOutput>
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(ServiceAndContextConfigurationProvider<TMapInput,TMapOutput>));
+
+        private readonly Dictionary<string, Tuple<IConfiguration, IConfiguration>> _configurationProvider;
+        private readonly ISet<string> _failedEvaluators;
+        private readonly ISet<string> _submittedEvaluators; 
+        private readonly object _lock;
+        private readonly int _numNodes;
+        private int _assignedPartitionDescriptors;
+        private readonly IGroupCommDriver _groupCommDriver;
+        private readonly ConfigurationManager _configurationManager;
+        private readonly IConfiguration _tcpPortProviderConfig;
+        private readonly Stack<IPartitionDescriptor> _partitionDescriptors;
+
+        internal ServiceAndContextConfigurationProvider(int numNodes, IGroupCommDriver groupCommDriver,
+            ConfigurationManager configurationManager, IConfiguration tcpPortProviderConfig, Stack<IPartitionDescriptor> partitionDescriptors)
+        {
+            _configurationProvider = new Dictionary<string, Tuple<IConfiguration, IConfiguration>>();
+            _failedEvaluators = new HashSet<string>();
+            _submittedEvaluators = new HashSet<string>();
+            _numNodes = numNodes;
+            _groupCommDriver = groupCommDriver;
+            _configurationManager = configurationManager;
+            _tcpPortProviderConfig = tcpPortProviderConfig;
+            _assignedPartitionDescriptors = 0;
+            _partitionDescriptors = partitionDescriptors;
+            _lock = new object();
+        }
+
+        /// <summary>
+        /// Handles failed evaluator. Moves the id from 
+        /// submitted evaluator to failed evaluator
+        /// </summary>
+        /// <param name="evaluatorId"></param>
+        internal void EvaluatorFailed(string evaluatorId)
+        {
+            lock (_lock)
+            {
+                if (!_submittedEvaluators.Contains(evaluatorId))
+                {
+                    Exceptions.Throw(new Exception("Failed evaluator was never submitted"), Logger);
+                }
+
+                _failedEvaluators.Add(evaluatorId);
+                _submittedEvaluators.Remove(evaluatorId);
+            }
+        }
+
+        /// <summary>
+        /// Gives configuration for next evaluator either from failed 
+        /// evaluator or new configuration
+        /// </summary>
+        /// <param name="evaluatorId"></param>
+        /// <returns></returns>
```

Comment:
Please document the return values here. `IConfiguration` could be anything. I only found which is which by looking at the usage of this method. Also, consider not using `Tuple` but instead a custom struct `ContextAndServiceConfiguration` or such.